File: java/src/ch1/Ex125.java
package ch1;

import java.util.Date;
import java.util.Random;

public class Ex125 {
    public static void main(String[] args) {
        System.out.println(runtime());
        searchForPrimes(1000, 1020);
        searchForPrimes(10000, 10040);
        searchForPrimes(100000, 100080);
        System.out.println("done");
    }

    public static boolean odd(long n) {
        return !divides(2, n);
    }

    public static long square(long n) {
        return n * n;
    }

    public static boolean divides(long a, long b) {
        return 0 == b % a;
    }

    public static boolean prime(int n) {
        return fast_prime(n, 3);
    }

    public static boolean fast_prime(int n, int times) {
        if (times == 0)
            return true;
        else if (fermat_test(n))
            return fast_prime(n, times-1);
        else
            return false;
    }

    @FunctionalInterface
    interface TryFermat { boolean apply(long n); }

    public static boolean fermat_test(int n) {
        TryFermat f = (long a) -> a == new ExpMod(a, n, n).value();
        return f.apply(1 + random(n-1));
    }

    public static int random(int limit) {
        return new Random().nextInt(limit);
    }

    public static long runtime() {
        return new Date().getTime();
    }

    public static void reportPrime(long elapsedTime) {
        System.out.print(" *** ");
        System.out.println(elapsedTime);;
    }

    public static void reportNonPrime() {
        //do nothing System.out.println("  ");
    }

    public static void startPrimeTest(int n, long startTime) {
        if (prime(n)) {
            reportPrime(runtime() - startTime);
        } else {
            reportNonPrime();
        }
    }

    public static void timedPrimeTest(int n) {
        System.out.println();
        System.out.print(n);
        startPrimeTest(n, runtime());
    }

    public static void searchForPrimes(int start, int end) {
        searchHelper(odd(start) ? start : ++start, end);
        System.out.println("finished");
    }

    public static void searchHelper(int n, int end) {
        if (n <= end) {
            timedPrimeTest(n);
            searchHelper(n+2, end);
        } else {
            System.out.println(" ");
        }
    }

    static class ExpMod implements Value<Long> {
        long base;
        long exp;
        long m;
        ExpMod(long base, long exp, long m) {
            this.base = base;
            this.exp = exp;
            this.m = m;
        }

        @Override
        public Long value() {
            return calc(this.base, this.exp, this.m);
        }

        // This breaks the algo as in other languages
        long calc(long base, long exp, long m) {
            return fast_expt(base, exp) % m;
        }

        long fast_expt(long base, long exp) {
            if (exp == 0) {
                return 1;
            } else if (divides(2, exp)) {
                return fast_expt(base, exp/2);
            } else {
                return (base * fast_expt(base, exp-1));
            }
        }
    }
}

File: ruby/ch1/ex125.rb
module Ex124

def divides(a, b)
  0 == b % a
end

def square(n)
  n * n
end

def find_divisor(n, test)
  if test * test > n
    n
  elsif divides(test, n)
    test
  else
    find_divisor(n, test + 1)
  end
end 

def smallest_divisor(n)
  find_divisor(n, 2)
end

def is_prime(n)
  # n == smallest_divisor(n)
  fast_prime(n, 3)
end

def fast_prime(n, times)
  if times == 0
    true
  elsif fermat_test(n)
    fast_prime(n, times-1)
  else
    false
  end
end

def fermat_test(n)
  try_it = lambda { |a| a == expmod(a, n, n) }
  try_it.call(1 + rand(n-1))
end

def expmod(base, exp, m)
  fast_expt(base, exp) % m
end

def fast_expt(base, exp)
  if exp == 0
    1
  elsif divides(2, exp)
    square(fast_expt(base, exp/2))
  else
    base * fast_expt(base, exp-1)
  end
end

def report_prime(n)
  result = is_prime(n) ? "***" : ""
  print("#{n} #{result}\n")
end

def search_helper(n, finish)
  if n <= finish
    report_prime(n)
    search_helper(n + 2, finish)
  else
    print("finished\n\n")
  end
end

def search_for_primes(start, finish)
  search_helper(divides(2, start) ? start + 1 : start, finish)
end

end

include Ex124
print "Testing\n"

search_for_primes(1000, 1020)
search_for_primes(10000, 10040)
search_for_primes(100000, 100080)

print "done", "\n"
